src/v/cluster/log_eviction_stm.cc

```diff
@@ -42,28 +42,53 @@ log_eviction_stm::log_eviction_stm(
   raft::consensus* raft, ss::logger& logger, storage::kvstore& kvstore)
   : base_t("log_eviction_stm.snapshot", logger, raft, kvstore) {}
 
+void log_eviction_stm::maybe_report_watchdog_timer_error() {
+    // sample max removable offset
+    const auto now = model::timeout_clock::now();
+    const auto max_removable
+      = _raft->log()->stm_manager()->max_removable_local_log_offset();
+    const auto timeout
+      = config::shard_local_cfg().partition_manager_shutdown_watchdog_timeout();
+    /**
+     * Update max removable offset if it has changed since the last time
+     */
+    if (max_removable > _loop_state.last_max_removable_offset) {
+        _loop_state.last_max_removable_offset = max_removable;
+        _loop_state.last_max_removable_offset_update = now;
+    }
+    /**
+     * Detect if max removable offset has not changed since the last time.
+     */
+    if (max_removable < _raft->committed_offset()) {
+        if (_loop_state.last_max_removable_offset_update + timeout < now) {
+            vlog(
+              _log.error,
+              "max removable offset is not advancing. current state: {}",
+              _loop_state);
+            _loop_state.max_removable_offset_update_stalls++;
```

Comment:
I was wondering if this reporting works fully as intended.. this requires the timer to be armed + fired when it is stuck. If the state = waiting_for_events, we don't rearm the timer (L74) and the storage subsystem doesn't request eviction if max_removable_local_offset is stuck, so this condition may not execute, ever? (hope I'm not missing something obvious).

Reply:
actually i will take it out from the log_eviction stm